internal/controller/ansibleRun/ansibleRun.go

```diff
@@ -527,3 +554,164 @@ func addBehaviorVars(pc *v1alpha1.ProviderConfig) map[string]string {
+			},
+			Spec: coordinationv1.LeaseSpec{
+				HolderIdentity:       &c.replicaID,
+				RenewTime:            &metav1.MicroTime{Time: time.Now()},
+				LeaseDurationSeconds: leaseDurationSeconds,
+			},
+		}
+		if err := kube.Create(ctx, lease); err != nil {
+			return err
+		}
+		c.logger.Debug("created lease", "lease", lease)
+		return nil
+	}
+
+	// Check if the lease is held by another replica and is not expired
+	if lease.Spec.HolderIdentity != nil && *lease.Spec.HolderIdentity != c.replicaID {
+		if lease.Spec.RenewTime != nil && time.Since(lease.Spec.RenewTime.Time) < time.Duration(*lease.Spec.LeaseDurationSeconds)*time.Second {
+			// Lease is held by another replica and is not expired
+			return fmt.Errorf("lease is still held by %s", *lease.Spec.HolderIdentity)
+		}
+	}
+
+	// Update the lease to acquire it
+	lease.Spec.HolderIdentity = pointer.String(c.replicaID)
+	lease.Spec.RenewTime = &metav1.MicroTime{Time: time.Now()}
+	lease.Spec.LeaseDurationSeconds = leaseDurationSeconds
+	if err := kube.Update(ctx, lease); err != nil {
+		if kerrors.IsConflict(err) {
+			// Another replica updated the lease concurrently, retry
+			return err
+		}
+		return fmt.Errorf("failed to update lease: %w", err)
+	}
+
+	c.logger.Debug("updated lease", "lease", lease)
+	return nil
+}
+
+// Finds an available shard and acquires a lease for it. Will attempt to obtain one indefinitely.
+// This will also start a background go-routine to renew the lease continiously and release it when the process receives a shutdown signal
+func (c *connector) acquireAndHoldShard(o controller.Options, s SetupOptions) (int, error) {
+	ctx := context.Background()
+	acquiredLease := false
+	currentShard := -1
+
+	cfg := ctrl.GetConfigOrDie()
+	kube, err := client.New(cfg, client.Options{})
+	if err != nil {
+		return currentShard, err
+	}
+
+	for !acquiredLease {
+		for i := 0; i < s.ReplicasCount; i++ {
+			if err := c.acquireLease(ctx, kube, i); err == nil {
+				currentShard = i
+				o.Logger.Debug("acquired lease", "id", i)
+				acquiredLease = true
+				go func() {
+					sigHandler := ctrl.SetupSignalHandler()
+
+					for {
+						select {
+						case <-time.After(leaseRenewalInterval):
+							if err := c.acquireLease(ctx, kube, i); err != nil {
+								o.Logger.Info("failed to renew lease", "id", i, "err", err)
+								panic(err)
+							} else {
+								o.Logger.Debug("renewed lease", "id", i)
+							}
+						case <-sigHandler.Done():
+							o.Logger.Info("controller is shutting down, releasing lease")
+							if err := c.releaseLease(ctx, kube, i); err != nil {
+								o.Logger.Info("failed to release lease", "lease", err)
+							}
+							o.Logger.Debug("released lease")
+							s.ProgramCancel()
+							return
+						}
+					}
+				}()
+				// Lease is acquired and background goroutine started for renewal, we can safely break to return the current shard
+				break
+			} else {
+				o.Logger.Debug("cannot acquire lease", "id", i, "err", err)
```

Comment:
```suggestion
				o.Logger.Error(err, "cannot acquire lease", "id", i)
```

Reply:
The logger used in there doesn't support `.Error`